### DISCUSSION TITLE
Optimizing USB Capture on a Raspberry Pi 4B

### DISCUSSION THREAD

**Discussion by d8ahazard** [emgucv/emgucv#564] [Q&A]
Using release 4.5.1.4349 from Nuget in c# on a raspberry pi 4B.

My USB input device is a generic HDMI capture card, which I'm using to do some magick and drive various lighting devices.

However, I'm running into a snag where the fastest I can get out of it is 15fps. On windows, using a beast of a machine and D-show, it pulls ~30FPS consistently.

I've done some basic profiling to try and figure out why, and it seems as if the bottleneck is definitely in the VideoCapture.Read() function. I've pulled out all of my other processing, and just tried creating a new mat(), reading to it, and then disposing it and counting the frames per second (using a StopWatch()).

I did some Goooogling on the subject, and it seems as if there *might be* a way to leverage FFMPEG and/or the H264 codec to speed up video capture, but then again, it might just be a limitation of some hardware. Either way, the bits I've seen are for openCV, and so I've been hesitant to try them, aside from what I mention below.

Doing a lookiup with v4l2-ctl --list-formats-ext, I can see that my device has a yuv profile for 640x480@60fps, which is the frame size I'm fine with pulling. In my app, I'm only setting CapProp.Width and Height normally, but I've messed around with changing the FPS, FourCC, and API, but none of those have given me any difference in frame rate.

Any experts out there who could help me get a faster capture speed? I can send to other devices at upwards of 30FPS, so I just need to get the video cap up to par as well and then I'll be good to go. :D

**Comment by emgucv** [comment]
It is probably a question that should be asked on the Open CV repo. 

Please also check if you are using V4L backend when capturing. 

When you change the VideoCapture property, there is a "bool" flag that is returned, check to see if the returned flag is "true":
https://www.emgu.com/wiki/files/4.5.3/document/html/c5a43e35-529e-2cb8-6605-86a21a775549.htm 

Also, depends on the backend, some property are the "preferred" capture property and is not a "guranteed" value. See the "notes" in OpenCV's documentation:
https://docs.opencv.org/4.5.3/d8/dfe/classcv_1_1VideoCapture.html#a8c6d8c2d37505b5ca61ffd4bb54e9a7c

> Even if it returns true this doesn't ensure that the property value has been accepted by the capture device.

**Reply by d8ahazard** [reply]
> It is probably a question that should be asked on the Open CV repo.
> 
> Please also check if you are using V4L backend when capturing.
> 
> When you change the VideoCapture property, there is a "bool" flag that is returned, check to see if the returned flag is "true":
> https://www.emgu.com/wiki/files/4.5.3/document/html/c5a43e35-529e-2cb8-6605-86a21a775549.htm
> 
> Also, depends on the backend, some property are the "preferred" capture property and is not a "guranteed" value. See the "notes" in OpenCV's documentation:
> https://docs.opencv.org/4.5.3/d8/dfe/classcv_1_1VideoCapture.html#a8c6d8c2d37505b5ca61ffd4bb54e9a7c
> 
> > Even if it returns true this doesn't ensure that the property value has been accepted by the capture device.

Actually, just getting a different insight into the issue was enough to help.

After using GetProp to read the FourCC and FPS (and backend), I noticed it was defaulting to YUYV, whereas MJPEG has the target fps/size combo I was looking for. By swtting the FourCC to MJPEG and manually setting the capProps for the size and frame rate, I can now pull ~30 fps on a pi, which is good enough for me. 

Thanks!
